### PR TITLE
feat/plugins

### PR DESCRIPTION
**Pull request by edbzn**
Following the #71 discussion, this working draft implements the plugin hook system used by `semantic-release`.

- Load plugins from workspace def
- Handle plugins options
- Run publish hook
- *todo*: implement other hooks
- *todo*: handle errors
- *todo*: validate inputs
- *todo*: pass right context object to plugin

### CODE REVIEW

File: packages/semver/src/builders/version/builder.spec.ts [hunk 1]
```diff
@@ -141,6 +155,84 @@ describe('@jscutlery/semver:version', () => {
     expect(output).toEqual(expect.objectContaining({ success: false }));
   });
 
+  describe('Plugins', () => {
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/builder.spec.ts:158]: @edbzn maybe we should move this describe in a distinct file.
There is already too much mocking in the parent describe `@jscutlery/semver:version` 😅.
This could be a dedicated suite where we would only mock `tryBump` and an internal `version` function that calls `versionProject` or `versionWorkspace`.

File: packages/semver/src/builders/version/builder.spec.ts [hunk 2]
```diff
@@ -141,6 +155,84 @@ describe('@jscutlery/semver:version', () => {
...
+    /* eslint-disable @typescript-eslint/no-var-requires */
+    const { publish: npmPublish } = require('@custom-plugin/npm');
+    const { publish: githubPublish } = require('@custom-plugin/github');
+    const {
+      publish: semanticPublish,
+      addChannel: semanticAddChannel,
+    } = require('@semantic-release/npm');
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/builder.spec.ts:165]: We can move this to the top of the file. It doesn't have to be in the describe.

File: packages/semver/src/builders/version/builder.spec.ts [hunk 3]
```diff
@@ -141,6 +155,84 @@ describe('@jscutlery/semver:version', () => {
...
+      expect(output).toEqual(expect.objectContaining({ success: true }));
+    });
+
+    it('should handle plugin configuration', async () => {
+      const output = await runBuilder(
+        {
+          ...options,
+          plugins: [
+            '@custom-plugin/npm',
+            ['@custom-plugin/github', { remoteUrl: 'remote' }],
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/builder.spec.ts:205]: As mentioned in `plugin.ts`, this could be:
```suggestion
            '@custom-plugin/npm',
            { module: '.../github', options: {remoteUrl: 'remote'}},
```

File: packages/semver/src/builders/version/builder.spec.ts [hunk 4]
```diff
@@ -141,6 +155,84 @@ describe('@jscutlery/semver:version', () => {
...
+          plugins: ['@semantic-release/npm'],
+        },
+        context
+      ).toPromise();
+
+      /* Ensure the adapter calls the semantic-release hooks. */
+      expect(semanticPublish).toBeCalled();
+      expect(semanticAddChannel).toBeCalled();
+      expect(output).toEqual(expect.objectContaining({ success: true }));
+    });
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/builder.spec.ts:233]: I would move this to another test that tests the `SemanticReleasePluginAdapter`.

File: packages/semver/src/builders/version/builder.ts [hunk 5]
```diff
@@ -63,7 +66,8 @@ export function runBuilder(
 
       return concat(
         runStandardVersion$,
-        ...(push && dryRun === false ? [pushToGitRemote$] : [])
+        ...(push && dryRun === false ? [pushToGitRemote$] : []),
+        pluginHandler.publish() // @todo handle dryRun
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/builder.ts:70]: Shouldn't we `defer(() => ...publish())`? 
We might need a test that makes sure that `publish` is only called after `version` finishes and succeeds.

File: packages/semver/src/builders/version/plugin-adapter.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+export interface SemanticReleasePlugin extends Plugin {
+  addChannel?(options: PluginOptions): Promise<unknown>;
+}
+
+export class SemanticReleasePluginAdapter implements Plugin {
+  constructor(private _plugin: SemanticReleasePlugin) {}
+
+  publish(options: PluginOptions) {
+    this._plugin.addChannel(options);
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/plugin-adapter.ts:13]: I think we should `concat` as  `addChannel` & `publish` are async.

File: packages/semver/src/builders/version/plugin-adapter.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    return this._plugin.publish(options);
+  }
+}
+
+export function _adapt(pluginName: string, plugin: Plugin): Plugin {
+  return SUPPORTED_SEMANTIC_RELEASE_PLUGINS.includes(pluginName)
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/plugin-adapter.ts:19]: 🤔 oh! I didn't think about a whitelist. How do we make sure that we are not calling a `semantic-release` plugin? Well, in the worst case, it will just crash 😊
I was imagining something that would detect if it's a `semantic-release` plugin or not.
By the way, as both `semver` and `semantic-release` plugins have a `publish` function but with a different signature, we can't detect the plugin using duck typing.

Maybe we could provide a `@jscutlery/semver-plugin-utils` util with a `createSemverPlugin` function that returns an object with some magic string in it: `{publish() {}, name: 'my-plugin', type: '@jscutlery/semver-plugin'}`.
We can check the `@jscutlery/semver-plugin` magic string.

This is strict but cumbersome as plugin developers will have to depend on `@jscutlery/semver-plugin-utils`... which might be cool as it could contain other helpers in the future.

If we don't do this from the beginning, we will have some plain object plugins in the wild and we wouldn't be able to verify the magic string as it would be a breaking change.

File: packages/semver/src/builders/version/plugin-handler.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  publish(): Observable<unknown> {
+    return this._run('publish');
+  }
+
+  private _run(hook: keyof Plugin): Observable<unknown> {
+    return defer(() => {
+      const plugins = this._plugins;
+      return Promise.all(
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/plugin-handler.ts:20]: I think we should run plugins sequentially as some plugins might have a side effect on others.

File: packages/semver/src/builders/version/plugin.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { JsonObject } from '@angular-devkit/core';
+
+export interface Plugin {
+  publish?(options: PluginOptions): Promise<unknown>;
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/plugin.ts:4]: Promise or Observable or any of them? 😊

**Review by yjaaidi — COMMENTED** [review]
This is exciting!
I added some technical details and some more conceptual discussions like the plugin detection with magic strings.

File: packages/semver/src/builders/version/builder.spec.ts (hunk 1 above)
Comment (edbzn) [reply packages/semver/src/builders/version/builder.spec.ts]: I moved the tests in a dedicated plugin.spec.ts file, it tests both the handler and the adapter at the same with an integration test, maybe this could be improved.

File: packages/semver/src/builders/version/builder.ts (hunk 5 above)
Comment (edbzn) [reply packages/semver/src/builders/version/builder.ts:72]: The publish function already defer it result. I agree, I will add a test for this.

File: packages/semver/src/builders/version/plugin-adapter.ts (hunk 6 above)
Comment (edbzn) [reply packages/semver/src/builders/version/plugin-adapter.ts]: Temporary fixed using async / await, I will improve this by using `concat` when Observable results will be handled.

File: packages/semver/src/builders/version/plugin-adapter.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  }
+}
+
+export function _isSemanticPlugin(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  plugin: any
+): plugin is RawSemanticReleasePlugin {
+  return (
+    typeof plugin.publish === 'function' &&
+    typeof plugin.verifyCondition === 'function'
```
Comment (yjaaidi) [inline packages/semver/src/builders/version/plugin-adapter.ts]: I think that all functions are optional. Apparently `analyzeCommits` is required but I think that what the documentation means is that at least one of the plugins should have the function.
TL;DR: There could be some plugins that have neither `publish` nor `verifyCondition`. https://github.com/semantic-release/semantic-release/blob/master/docs/usage/plugins.md

Comment (edbzn) [reply packages/semver/src/builders/version/plugin-adapter.ts]: Yep I was not sure about it, there's no easy way to do duck typing here. Maybe we can check if one hook exist at least?

Comment (yjaaidi) [reply packages/semver/src/builders/version/plugin-adapter.ts]: this sounds perfect 👌
